src/codegen/doc/format.rs

```diff
@@ -54,121 +60,352 @@ fn get_language<'a>(entry: &'a str, out: &mut String) -> &'a str {
+        .analysis
+        .constants
+        .iter()
+        .find(|c| c.glib_name == symbol)
+    {
+        Some(gen_const_doc_link(const_info))
+    } else {
+        warn!("Constant/Flag variant/Enum member `{}` not found", symbol);
+        None
+    }
+}
+
+// A list of types that are automatically ignored by the `find_type` function
+const IGNORED_C_TYPES: [&str; 6] = [
+    "gconstpointer",
+    "guint16",
+    "guint",
+    "gunicode",
+    "gchararray",
+    "GList",
+];
+/// either an object/interface, record, enum or a flag
+fn find_type(type_: &str, env: &Env, in_type: Option<&TypeId>) -> Option<String> {
+    if IGNORED_C_TYPES.contains(&type_) {
+        return None;
+    }
+
+    let type_id = if let Some(obj) = env.analysis.objects.values().find(|o| o.c_type == type_) {
+        Some(obj.type_id)
+    } else if let Some(record) = env
+        .analysis
+        .records
+        .values()
+        .find(|r| r.type_(&env.library).c_type == type_)
+    {
+        Some(record.type_id)
+    } else if let Some(enum_) = env
+        .analysis
+        .enumerations
+        .iter()
+        .find(|e| e.type_(&env.library).c_type == type_)
+    {
+        Some(enum_.type_id)
+    } else if let Some(flag) = env
+        .analysis
+        .flags
+        .iter()
+        .find(|f| f.type_(&env.library).c_type == type_)
+    {
+        Some(flag.type_id)
+    } else if let Some(constant_or_variant) = find_constant_or_variant(type_, env, in_type) {
+        warn!(
+            "`{}` matches a constant or enum-variant and should be prefixed with `%` instead of `#` (reserved for types)",
+            type_
+        );
+        return Some(constant_or_variant);
+    } else {
+        warn!("Object/Interface/Record `{}` not found", type_);
+        None
+    };
+
+    type_id.map(|ty| gen_symbol_doc_link(ty, env))
+}
+
+fn find_method_or_function_by_ctype(
+    c_type: Option<&str>,
+    name: &str,
+    env: &Env,
+    in_type: Option<&TypeId>,
+) -> Option<String> {
+    find_method_or_function(
+        name,
+        env,
+        in_type,
+        |f| f.glib_name == name,
+        |o| c_type.map_or(true, |t| o.c_type == t),
+        |r| c_type.map_or(true, |t| r.type_(&env.library).c_type == t),
+    )
+}
+
+// A list of function names that are ignored when warning about a "not found function"
+const IGNORE_C_WARNING_FUNCS: [&str; 6] = [
+    "g_object_unref",
+    "g_object_ref",
+    "g_free",
+    "g_list_free",
+    "g_strfreev",
+    "printf",
+];
+/// Find a function in all the possible items, if not found return the original name surrounded with backticks
+/// A function can either be a struct/interface/record method, a global function or maybe a virtual function
+///
+/// This function is generic so it can be de-duplicated between a
+/// - "find_method_or_function_by_ctype" where the object/records are looked by their C name
+/// - "find_method_or_function_by_name" where the object/records are looked by their name
```

Comment:
Did `find_method_or_function_by_name` disappear again during some refactor?

Reply:
Btw, ``[`find_method_or_function_by_ctype`]`` :wink:

Reply:
Nop i refactored it a bit to reuse it for gi-docgen

Reply:
Oh, this sits in `gi_docgen`. Yeah, use intradoc-links to clarify that :)

Reply:
![image](https://user-images.githubusercontent.com/2325264/120071687-b934c900-c090-11eb-94d1-1a237a55d78d.png)

^^

Reply:
ah my bad. I forgot we are doing all this work to make docs better... haha